### PR TITLE
feat(#2543): added HmOptionalTest

### PR DESCRIPTION
**Pull request by Yanich96**
HmOptionalTest was added.

---

## PR-Codex overview
This PR focuses on updating the `g4s8/pdd-action` dependency to `volodya-lombrozo/pdd-action` in the GitHub Actions workflow. 

### Detailed summary
- Updated the `g4s8/pdd-action` dependency to `volodya-lombrozo/pdd-action` in the GitHub Actions workflow.

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

**Comment by maxonfjvipon** [comment]
@Yanich96 wait for https://github.com/g4s8/pdd-action/issues/9

**Comment by volodya-lombrozo** [comment]
@Yanich96 You can replace `g4s8/pdd-action@master` -> `volodya-lombrozo/pdd-action@master` in [this](https://github.com/objectionary/eo/blob/master/.github/workflows/pdd.yml) file and it should help.

**Comment by maxonfjvipon** [comment]
@volodya-lombrozo the repo was transferred to https://github.com/maxonfjvipon/pdd-action, so it would be nice if you make a pull request

**Comment by Yanich96** [comment]
@volodya-lombrozo thanks =) this works.

**Comment by Yanich96** [comment]
@maxonfjvipon check my code, please

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+    @BeforeEach
+    void setUp() throws Exception {
+        this.size = 100;
+        this.sample = "file.txt";
+    }
+
+    @ParameterizedTest
+    @CsvSource({"true", "false"})
+    void savesIfFileNotExist(final boolean rewrite, @TempDir final Path dir) throws IOException {
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:69]: @Yanich96 I think the right english name for the test would be `savesIfFileDoesNotExist`

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        final HmOptional optional = new HmOptional(new HmBase(dir), rewrite);
+        final Path file = Paths.get(this.sample);
+        final String content = new UncheckedText(new Randomized(this.size)).asString();
+        optional.save(content, file);
+        MatcherAssert.assertThat(
+            new UncheckedText(new TextOf(dir.resolve(file))).asString(),
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:75]: @Yanich96 let's add human readable assertion message here

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            Matchers.is(content)
+        );
+    }
+
+    @Test
+    void savesIfFileExistAndRewriteTrue(@TempDir final Path dir) throws IOException {
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:81]: @Yanich96 I think the right english name for the test would be `savesIfFileExistsAndRewriteTrue`

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        base.save(first, file);
+        final HmOptional optional = new HmOptional(base, true);
+        final String second = new UncheckedText(new Randomized(this.size)).asString();
+        optional.save(second, file);
+        MatcherAssert.assertThat(
+            new UncheckedText(new TextOf(dir.resolve(file))).asString(),
+            Matchers.not(first)
+        );
+        MatcherAssert.assertThat(
+            new UncheckedText(new TextOf(dir.resolve(file))).asString(),
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:94]: @Yanich96 Let's add human readable assertion message here

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    @Test
+    void savesIfFileExistAndRewriteTrue(@TempDir final Path dir) throws IOException {
+        final String first = new UncheckedText(new Randomized(this.size)).asString();
+        final Path file = Paths.get(this.sample);
+        final HmBase base = new HmBase(dir);
+        base.save(first, file);
+        final HmOptional optional = new HmOptional(base, true);
+        final String second = new UncheckedText(new Randomized(this.size)).asString();
+        optional.save(second, file);
+        MatcherAssert.assertThat(
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:89]: @Yanich96 I think this assertion can be deleted, the below one is enough for the test. They exclude each other

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        );
+        MatcherAssert.assertThat(
+            new UncheckedText(new TextOf(dir.resolve(file))).asString(),
+            Matchers.is(second)
+        );
+    }
+
+    @Test
+    @CaptureLogs
+    void savesIfFileExistAndRewriteFalse(
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:101]: @Yanich96 I think the right english name for the test would be `savesIfFileExistsAndRewriteFalse`

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+        base.save(first, file);
+        final HmOptional optional = new HmOptional(base, false);
+        final String second = new UncheckedText(new Randomized(this.size)).asString();
+        optional.save(second, file);
+        final Path absolute = dir.resolve(file);
+        MatcherAssert.assertThat(
+            new UncheckedText(new TextOf(absolute)).asString(),
+            Matchers.is(first)
+        );
+        MatcherAssert.assertThat(
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:116]: @Yanich96 I think this assertion can be deleted, the above one is enough for the test. They exclude each other

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        final String first = new UncheckedText(new Randomized(this.size)).asString();
+        final Path file = Paths.get(this.sample);
+        final HmBase base = new HmBase(dir);
+        base.save(first, file);
+        final HmOptional optional = new HmOptional(base, false);
+        final String second = new UncheckedText(new Randomized(this.size)).asString();
+        optional.save(second, file);
+        final Path absolute = dir.resolve(file);
+        MatcherAssert.assertThat(
+            new UncheckedText(new TextOf(absolute)).asString(),
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:113]: @Yanich96 Let's add human readable assertion message here

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            )
+        );
+    }
+
+    @Test
+    void exists(@TempDir final Path dir) throws IOException {
+        final Path file = Paths.get(this.sample);
+        Files.write(dir.resolve(file), "any content".getBytes());
+        MatcherAssert.assertThat(
+            new HmOptional(new HmBase(dir), true).exists(file),
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:134]: @Yanich96 Let's add human readable assertion message here

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    @Test
+    void absolutes(@TempDir final Path dir) throws IOException {
+        final Path file = Paths.get(this.sample);
+        final HmBase base = new HmBase(dir);
+        base.save("", file);
+        final Path absolute = dir.resolve(file);
+        Assertions.assertEquals(
+            absolute,
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:146]: @Yanich96 Let's add human readable assertion message here

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+        );
+    }
+
+    @ParameterizedTest
+    @CsvSource({"file.txt", "a/file.txt", "a/b/file.txt"})
+    void checksOnlyRelative(final Path file, @TempDir final Path dir) throws IOException {
+        final HmBase base = new HmBase(dir);
+        base.save("", file);
+        Assertions.assertEquals(
+            file,
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:157]: @Yanich96 Let's add human readable assertion message here

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    @Test
+    void loads(@TempDir final Path dir) throws IOException {
+        final HmBase base = new HmBase(dir);
+        final Path file = Paths.get(this.sample);
+        final String text = "Hello World";
+        base.save(text, file);
+        Assertions.assertEquals(
+            text,
```
Comment (maxonfjvipon) [inline eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:169]: @Yanich96 Let's add human readable assertion message here

**Review by maxonfjvipon — CHANGES_REQUESTED** [review]
@Yanich96 well done, just a few small comments from my side

File: eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java (hunk 2 above)
Comment (Yanich96) [reply eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:75]: I wanted to add a message but "MatcherAssert.assertThat" does not accept messages

Comment (maxonfjvipon) [reply eo-maven-plugin/src/test/java/org/eolang/maven/util/HmOptionalTest.java:75]: @Yanich96 It actually does:
image

**Comment by Yanich96** [comment]
@maxonfjvipon I have done

**Review by maxonfjvipon — APPROVED** [review]
@Yanich96 LGTM! Thanks

**Comment by Yanich96** [comment]
@yegor256 approve please

**Pull request merged**